### ISSUE TITLE
[v4] Backdrop opacity

### ISSUE THREAD

**Issue by pjc0247** [ionic-team/ionic#16446]
![image](https://user-images.githubusercontent.com/1961216/48979316-9199ca00-f0fc-11e8-936d-d125166abaa0.png)

There's already `backdrop-color` variable for customization. but `opacity` is hard-coded in the `element.style`.

It would be great to give a change to modify backdrop opacity with global variable.

**Comment by brandyscarney** [comment]
Thanks for the issue! So the issue here is actually that the opacity is added to the backdrop in the animation (enter/leave) of the component. Which component are you looking to change the opacity on?

**Comment by pjc0247** [comment]
https://github.com/ionic-team/ionic/blob/master/core/src/components/popover/animations/md.enter.ts#L74

**Comment by bowenliang123** [comment]
how can i change the backdrop opacity of an overlay sidemenu? please help

**Comment by Frotty** [comment]
This problem existed for over 2 years now... https://github.com/ionic-team/ionic/issues/9105
Why can't we get a css variable for target opacity or something?

**Comment by Takkuz** [comment]
Same here... We've customers that need custom opacity (not interested in MD spec) and we're unable to give them this requirement. I was looking for a workaround withou success.

I think `--ion-backdrop-color` should define the opacity himself `rgba(0, 0, 0, .32)` granting user to override it easily.

**Comment by brandyscarney** [comment]
Please see my comment here: https://github.com/ionic-team/ionic/issues/17669#issuecomment-470243422

At the moment it can't be done using a CSS variable, but the opacity is set by the animation and it is possible to pass a custom animation. I mentioned there that we need to look into a way to set it easier, but it is possible. 

[Customizing animations via the config](https://ionicframework.com/docs/utilities/config#customizing-animations)

So where you see the following line you would change it from `0.32` to whatever number you'd like:

```
backdropAnimation.fromTo('opacity', 0.01, 0.32);
```

**Comment by cptflammin** [comment]
In your theme, workaround:

:root {
   ion-backdrop {
    opacity: 0.5 !important;
  }
}

**Comment by dani-inside** [comment]
This is a real issue for us, as we cannot change the opacity of the backdrop which conflicts with the desired designs. When changing it via SCSS it will break the fade animation. Any news on this?

**Comment by ShaneJohnsonCC** [comment]
This was working for me for a long time and now all of a sudden, it does nothing. Even if I try to change the opacity in inspect element, literally nothing happens. I can do anything else like change colors in inspect element. Does someone know what is going on?

`ion-modal ion-backdrop {
    opacity: 0.6 !important;
}`

**Comment by Frotty** [comment]
> This was working for me for a long time and now all of a sudden, it does nothing.

I shouldn't have worked. Perhaps you prevented the fade animation somehow via some side effect.
As mentioned before, the css value actually isn't used/respected in the animation and the css rule even if working, broke it.
You can see this happen on your page, where the backdrop has the desired 0.9 opacity for a split second before the animation starts.
As you can see from a quick GH-repo search, the animations all use different magic inline values for the opacity.
This is frankly terrible, I don't understand it and I don't think there is an elegant way to change these values.

![](https://i.imgur.com/xVtlquZ.jpg)

It's very disheartening to see simple issues like this be ignored or take forever to fix.

**Comment by ShaneJohnsonCC** [comment]
> It's very disheartening to see simple issues like this be ignored or take forever to fix.

I have only started building with Ionic since V3. So that is disappointing to hear. They usually do fix bugs pretty fast. But I guess this isn't really a "bug" per-say. I don't know... Hope they do make this more robust for us. Seems like a simple CSS4 variable added to an active class would be so much easier.

**Comment by ihadeed** [comment]
In case anyone here is still looking for a solution:

When creating a new loader, action-sheet, modal... etc, you have an option to provide `enterAnimation` and `leaveAnimation`. 

Those two properties take a callback function of the following signature:
```ts
((Animation: Animation, baseEl: any, opts?: any) => Promise<Animation>
```

I had some strange behaviour using that signature. In some cases I received `Animation` and `baseEl` correctly; then it gets called immediately after with only `baseEl` as the first param. See example below for a working callback function.

**Animation**: a class that you can create a new instance of and start building your own animation. 
This class is exported from a directory called `animation-old` so I assume that it's deprecated. There is another method called `createAnimation` which returns an `Animation` object. It seems to work fine. Here is an example of that function standalone:
```ts
import { createAnimation } from '@ionic/core';

...

// have your element ready
const someElement: HTMLElement = document.querySelector('...'); // or can be retrieved with Angular using ElementRef.nativeElement

// call createAnimation to get an animation instance
const animation = createAnimation();

// add the element to the the animation instance, so anything we do applies to it
animation.addElement(someElement);

// set the starting behavior / style
animation.beforeStyles({
  opacity: 0,
});

// set the duration and easing of animation
// we can chain all function calls (including all of the above)
animation.duration(250).easing('ease-in');

// play the animation
animation.play();
```

---

Here's an example of how you would use it for a custom loading animation.

The `baseEl` argument in the callback function is your base element depending on where you're using this `enterAnimation` callback. Which could be `HTMLIonLoadingElement`, `HTMLIonModalElement` ... etc

```ts
import { LoadingController } from '@ionic/angular';

...

constructor(private loadingCtrl: LoadingController) {}

...

showLoader() {
  this.loadingCtrl.create({
    // your options here
      enterAnimation: ((baseEl: HTMLIonLoadingElement) => {

        // create a backdrop animation object to set behaviour for the backdrop only
        const backdropAnimation = createAnimation()
          .addElement(baseEl.querySelector('ion-backdrop'))
          .fromTo('opacity', 0, 0.9)
          .duration(250)
          .easing('ease-in');

        // create a content animation to animate the loading content separately (we want to give it full opacity in this case)
        const contentAnimation = createAnimation()
          .addElement(baseEl.querySelector('.loading-wrapper'))
          .fromTo('opacity', 0, 1)
          .duration(250)
          .easing('ease-in');

        // merge them into one root animation and return that
        // you can optionally apply animations to the root component
        return createAnimation()
          .addAnimation([
            backdropAnimation,
            contentAnimation,
          ]);
      }) as any, // wrap the whole callback in brackets and apply `as any` since the signature doesn't match the interface
  });
}
```

Keep in mind that if you provide call back for `enterAnimation` or `leaveAnimation` you will be overriding the default enter or leave animations respectively. The default values for `z-index`, `opacity`, or other properties might prevent you from seeing the component completely until you define values for them in your animation using `from`, `to`, `fromTo`, `beforeStyles`, or `afterStyles`.

**Comment by ShaneJohnsonCC** [comment]
@ihadeed Thank you for the working example! However, it's not very smooth loading on my end. So I just set showBackdrop to false and then just did some css3 animation to fade in a background on the ion-modal. Just doing my own backdrop instead of using ionic's

**Comment by ihadeed** [comment]
@ShaneJohnsonCC I tried it with an `ion-loading` only, it worked well and it's smooth.

I think for a modal you might need additional animations to control the modal itself to set the position/height... etc. Might be worth looking into the source code to find the default animation that they use, and extend that manually (copy, paste, and modify/add as needed).

**Comment by ShaneJohnsonCC** [comment]
@ihadeed Makes perfect sense! Appreciate your help man

**Comment by ErlendEllingsen** [comment]
+1 on this

**Comment by sumbria** [comment]
Try this:

```
ion-popover {
  background: rgba(0, 0, 0, 0.7);
}
```

**Comment by brandyscarney** [comment]
Thanks for the feedback everyone. While we would love to fix every issue in a timely manner, there are over 700 issues and those with workarounds are not going to receive our immediate attention as we have critical issues without workarounds. We always welcome PRs in order to push these issues to the top of the list. 🙂 

I created a PR for this here that just needs review: https://github.com/ionic-team/ionic/pull/19533

I released a dev build if anyone would like to try it out: `4.11.0-dev.201910021700.08b22da`

**However** this is based on our `next` branch which is going to end up being a major release with breaking changes, any breaking change has a deprecation warning in latest 4.x, see the commits on the branch for a full list: https://github.com/ionic-team/ionic/commits/next

Thanks!!
